swift/extractor/visitors/VisitorBase.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <swift/AST/ASTVisitor.h>
+#include <swift/AST/TypeVisitor.h>
+
+#include "swift/extractor/SwiftDispatcher.h"
+
+namespace codeql {
+
+namespace detail {
+class DispatcherWrapper {
+ protected:
+  SwiftDispatcher& dispatcher_;
+
+ public:
+  // SwiftDispatcher should outlive this instance
+  DispatcherWrapper(SwiftDispatcher& dispatcher) : dispatcher_{dispatcher} {}
+};
+
+}  // namespace detail
+
+// we want to override the default swift visitor behaviour of chaining calls to immediate
+// superclasses by default and instead provide our own TBD default (using the exact type)
+#define DEFAULT(KIND, CLASS, PARENT) \
+  void visit##CLASS##KIND(swift::CLASS##KIND* e) { dispatcher_.emitUnknown(e); }
+
+// base class for our AST visitors, getting a SwiftDispatcher member and default emission for
+// unknown/TBD entities
+template <typename CrtpSubclass>
```

Comment:
What does `crtp` stand for? Is this for me somewhat cryptic name needed for the macro programming going on below?

Reply:
That's [Curiously recurring template pattern](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern), as proposed by @sashabu [here](https://github.com/github/codeql/pull/9144#discussion_r872473570).

Reply:
this was suggested by @sashabu and stands for the [curiously recurring template pattern](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern). It's probably better to put a comment about it, as also for me that have used this pattern a lot and am very familiar with its full nomenclautre, the acronym takes a bit to ring a bell.